lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp

```diff
@@ -4238,6 +4238,318 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
+                  rewriter.create<Torch::PrimLoopConditionOp>(
+                      binder.getLoc(), loopConditionTrue, ValueRange({count}));
+                }
+                count = countLoop.getResult(0);
+                rewriter.create<Torch::PrimLoopConditionOp>(
+                    binder.getLoc(), loopConditionTrue, ValueRange({count}));
+              }
+              count = skipLoop.getResult(0);
+              // insert count "tf" into output
+              Value countFloat = rewriter.create<Torch::AtenFloatScalarOp>(
+                  binder.getLoc(), count);
+              Value dataList = rewriter.create<Torch::PrimListConstructOp>(
+                  binder.getLoc(),
+                  rewriter.getType<Torch::ListType>(
+                      rewriter.getType<Torch::FloatType>()),
+                  SmallVector<Value>{countFloat});
+              Value cstDtype = rewriter.create<Torch::ConstantIntOp>(
+                  binder.getLoc(), rewriter.getI64IntegerAttr(
+                                       (int)torch_upstream::ScalarType::Float));
+              SmallVector<int64_t> countShape{1};
+              auto countType = rewriter.getType<Torch::ValueTensorType>(
+                  countShape, resultType.getOptionalDtype());
+              Value countTensor = rewriter.create<Torch::AtenTensorOp>(
+                  binder.getLoc(), countType, dataList, /*dtype=*/cstDtype,
+                  /*layout=*/none, /*requires_grad=*/cstFalse);
+
+              if (is_2d) {
+                // ex: output: [[0,0],[0,0]] ->
+                // outputSequence: [[0,0]] , countUnsqueezed: [[1]] ->
+                // outputSequence:
+                // [[0,1]] -> output: [[0,1],[0,0]]
+                Value countUnsqueezed = rewriter.create<Torch::AtenUnsqueezeOp>(
+                    binder.getLoc(),
+                    rewriter.getType<Torch::ValueTensorType>(
+                        llvm::SmallVector<int64_t>{1, 1},
+                        resultType.getDtype()),
+                    countTensor, zero);
+                Value insertStart = rewriter.create<Torch::ConstantIntOp>(
+                    binder.getLoc(),
+                    rewriter.getI64IntegerAttr(ngram_indexes[ngram_i]));
+                Value insertEnd = rewriter.create<Torch::AtenAddIntOp>(
+                    binder.getLoc(), insertStart, one);
+
+                SmallVector<int64_t> outputForBatchShape;
+                outputForBatchShape.push_back(1);
+                outputForBatchShape.push_back(resultShape[1]);
+                auto outputForBatchType =
+                    rewriter.getType<Torch::ValueTensorType>(
+                        outputForBatchShape, resultType.getOptionalDtype());
+                outputForBatch = rewriter.create<Torch::AtenSliceScatterOp>(
+                    binder.getLoc(), outputForBatchType, outputForBatch,
+                    countUnsqueezed,
+                    /*dim=*/one, insertStart, insertEnd, /*step=*/one);
```

Comment:
I don't think this last change really simplified as much as it should have.

The point in my comment earlier is that `outputForBatch` doesn't need to be 1-d or 2-d dependent. It's just one batch regardless of how many batches there are, so it should just be a tensor of shape [number of n-grams]. Once all is said and done, you can insert `outputForBatch` into the actual output tensor at the very end of the batch loop, but I feel like that (and maybe the initialization of `outputForBatch`, since it might be better to get it from the loop arg in 1-d rather than making a new tensor) should be the only place where 1/2d needs to affect the control flow inside any of these loop bodies.

Reply:
Ok made some changes